pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,427 @@
+            Furthermore, the QNode arguments *not* marked in ``argnum`` will not be
+            considered in this test and if they resemble encoded inputs, the entire
+            spectrum might be incorrect or the circuit might not even admit one.
+
+        The ``advanced_spectrum`` function works in all interfaces:
+
+        .. code-block:: python
+
+            import tensorflow as tf
+
+            dev = qml.device("default.qubit", wires=1)
+
+            @qml.qnode(dev, interface='tf')
+            def circuit(x):
+                qml.RX(0.4*x[0], wires=0)
+                qml.PhaseShift(x[1]*np.pi, wires=0)
+                return qml.expval(qml.PauliZ(wires=0))
+
+            x = tf.constant([1., 2.])
+            res = advanced_spectrum(circuit)(x)
+
+        >>> print(res)
+        {"x": {(0,): [-0.4, 0.0, 0.4], (1,): [-3.14159, 0.0, 3.14159]}}
+
+        Finally, compare ``advanced_spectrum`` with :func:`.pennylane.simple_spectrum`, using
+        the following circuit.
+
+        .. code-block:: python
+
+            dev = qml.device("default.qubit", wires=2)
+
+            @qml.qnode(dev)
+            def circuit(x, y, z):
+                qml.RX(0.5*x**2, wires=0, id="x")
+                qml.RY(2.3*y, wires=1, id="y0")
+                qml.CNOT(wires=[1,0])
+                qml.RY(z, wires=0, id="y1")
+                return qml.expval(qml.PauliZ(wires=0))
+
+        First, note that we assigned ``id``s to the gates for which we will use
+        ``simple_spectrum``. This allows us to choose these gates in the computation:
+
+        >>> x, y, z = 0.1, 0.2, 0.3
+        >>> simple_spec_fn = qml.fourier.simple_spectrum(circuit, encoding_gates=["x", "y0", "y1"])
+        >>> simple_spec = simple_spec_fn(x, y, z)
+        >>> for _id, spec in simple_spec.items():
+        ...     print(f"{_id}: {spec}")
+        x: [-1.0, 0, 1.0]
+        y0: [-1.0, 0, 1.0]
+        y1: [-1.0, 0, 1.0]
+
+        As we can see, the preprocessing in the QNode is not included in the simple spectrum.
+        In contrast, the output of ``advanced_spectrum`` is:
+
+        >>> adv_spec = qml.fourier.advanced_spectrum(circuit, encoding_args={"y", "z"})
+        >>> for _id, spec in adv_spec.items():
+        ...     print(f"{_id}: {spec}")
+        y: {(): [-2.3, 0.0, 2.3]}
+        z: {(): [-1.0, 0.0, 1.0]}
+
+        Note that the values of the output are dictionaries instead of the spectrum lists, that
+        they include the prefactors introduced by classical preprocessing, and
+        that we would not be able to compute the advanced spectrum for ``x`` because it is
+        preprocessed non-linearily in the gate ``qml.RX(0.5*x**2, wires=0, id="x")``.
+
+    """
+    # pylint: disable=too-many-branches,protected-access
+    validation_kwargs = validation_kwargs or {}
+    encoding_args, argnum = _process_ids(encoding_args, argnum, qnode)
+    atol = 10 ** (-decimals) if decimals is not None else 1e-10
```

Comment:
This is because in the end, we want not only comparisons (for which `atol` is the much better quantity, I agree) but we want to round frequencies, which is not defined for general float `atol`s, as far as I know.
If we were to rely on `atol`-based comparisons everywhere, we could not do set operations anymore but would have to do some `numpy`-based set-like operation that kicks out values similar up to `atol`.